travertino/declaration.py

```diff
@@ -50,6 +50,7 @@ def validate(self, value):
                 pass
         if value == 'none':
             value = None
+            return value
```

Comment:
Instead of doing `return value` here, try removing:

```
        if value == 'none':
             value = None
```

See if that gets the tests to pass.

Reply:
@rayrrr , thanks. I've already tried that, doesn't work.

When I comment out the two lines,
```
        if value == 'none':
             value = None
```

The test cases of 'test_multiple_choices', 'test_none', 'test_values' failed.

Reply:
Too bad. I'll give it some more thought too and see if I can help you get this done.

Reply:
@stantonxu you should be able to fix this without changing any existing tests by removing the `return value` line, and changing the line `if value == 'none':` to (pseudocode) `if value == 'none' and <type is Color, Integer, or Number>:`. Try that.

Reply:
The `validate` code is as below, so the `<type is Color, Integer, or Number>:` code won't hit.

```
    def validate(self, value):
        if self.default:
            if value is None:
                return None
        if self.string:
            try:
                return value.strip()
            except AttributeError:
                pass
        if self.integer:
            try:
                return int(value)
            except (ValueError, TypeError):
                pass
        if self.number:
            try:
                return float(value)
            except (ValueError, TypeError):
                pass
        if self.color:
            try:
                return color(value)
            except ValueError:
                pass
        if value == 'none':
            value = None
            return value
        for const in self.constants:
            if value == const:
                return const

        raise ValueError("'{0}' is not a valid initial value".format(value))

```

Reply:
What we want to do is change `'none'` to `None` for Color, Integer, and Number types. Let's do that _before_ the conditional checks for those types.

```
 def validate(self, value):
        if self.default:
            if value is None:
                return None
        if value == 'none' and (self.color or self.integer or self.number):
            value = None
        if self.string:
            try:
                return value.strip()
            except AttributeError:
                pass
        if self.integer:
            try:
                return int(value)
            except (ValueError, TypeError):
                pass
        if self.number:
            try:
                return float(value)
            except (ValueError, TypeError):
                pass
        if self.color:
            try:
                return color(value)
            except ValueError:
                pass
        for const in self.constants:
            if value == const:
                return const

        raise ValueError("'{0}' is not a valid initial value".format(value))
```

Also, notice the lines near the end of this block starting with `for const in self.constants`. It means that for this to avoid raising a `ValueError`, the value needs to appear in the `constants.py` file. We need to move some constants from Toga to Travertino and then import them from the latter to the former.

Reply:
@stantonxu once pybee/travertino#5 and pybee/toga#564 are merged, the code block above should work. Do you understand?

Reply:
Actually, the other PR's might not be needed, but I still think they'll make these apps more tidy. Try the code block as is and see what happens.

Reply:
@rayrrr thanks for the suggestion, however it doesn't work out. Here is the error from test,
```
======================================================================
ERROR: test_none (tests.test_choices.PropertyChoiceTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/xus12/git/pybee/travertino/travertino/declaration.py", line 181, in setter
    value = choices.validate(value)
  File "/Users/xus12/git/pybee/travertino/travertino/declaration.py", line 60, in validate
    raise ValueError("'{0}' is not a valid initial value".format(value))
ValueError: 'none' is not a valid initial value

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/xus12/git/pybee/travertino/tests/test_choices.py", line 46, in test_none
    obj.prop = 'none'
  File "/Users/xus12/git/pybee/travertino/travertino/declaration.py", line 184, in setter
    value, name, choices
ValueError: Invalid value 'none' for property 'prop'; Valid values are: none

======================================================================
ERROR: test_values (tests.test_choices.PropertyChoiceTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/xus12/git/pybee/travertino/travertino/declaration.py", line 181, in setter
    value = choices.validate(value)
  File "/Users/xus12/git/pybee/travertino/travertino/declaration.py", line 60, in validate
    raise ValueError("'{0}' is not a valid initial value".format(value))
ValueError: 'none' is not a valid initial value

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/xus12/git/pybee/travertino/tests/test_choices.py", line 260, in test_values
    obj.prop = 'none'
  File "/Users/xus12/git/pybee/travertino/travertino/declaration.py", line 184, in setter
    value, name, choices
ValueError: Invalid value 'none' for property 'prop'; Valid values are: a, b, none

----------------------------------------------------------------------
Ran 84 tests in 0.038s

FAILED (errors=2)
```